database/src/main/java/com/firebase/ui/database/FirebaseIndexArray.java

```diff
@@ -0,0 +1,150 @@
+
+import java.util.ArrayList;
+import java.util.List;
+
+class FirebaseIndexArray extends FirebaseArray implements ValueEventListener {
+    private static final String TAG = FirebaseIndexArray.class.getSimpleName();
+
+    private DatabaseReference mRef;
+    private List<DataSnapshot> mDataSnapshots = new ArrayList<>();
+    private OnChangedListener mListener;
+
+    FirebaseIndexArray(Query keyRef, DatabaseReference dataRef) {
+        super(keyRef);
+        mRef = dataRef;
+    }
+
+    @Override
+    public void cleanup() {
+        super.cleanup();
+        for (DataSnapshot snapshot : mDataSnapshots) {
+            snapshot.getRef().removeEventListener((ValueEventListener) this);
+        }
+    }
+
+    @Override
+    public int getCount() {
+        return mDataSnapshots.size();
+    }
+
+    @Override
+    public DataSnapshot getItem(int index) {
+        return mDataSnapshots.get(index);
+    }
+
+    private int getIndexForKey(String key) {
+        int dataCount = getCount();
+        int index = 0;
+        for (int keyIndex = 0; index < dataCount; keyIndex++) {
+            String superKey = super.getItem(keyIndex).getKey();
+            if (key.equals(superKey)) {
+                break;
+            } else if (getItem(index).getKey().equals(superKey)) {
+                index++;
+            }
+        }
+        return index;
+    }
+
+    private boolean isMatch(int index, String key) {
+        return index >= 0 && index < getCount() && getItem(index).getKey().equals(key);
+    }
+
+    @Override
+    public void onChildAdded(DataSnapshot keySnapshot, String previousChildKey) {
+        super.setOnChangedListener(null);
+        super.onChildAdded(keySnapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
+        mRef.child(keySnapshot.getKey()).addValueEventListener(this);
+    }
+
+    @Override
+    public void onDataChange(DataSnapshot snapshot) {
+        String key = snapshot.getKey();
+        int index = getIndexForKey(key);
+
+        if (snapshot.getValue() != null) {
+            if (!isMatch(index, key)) {
+                mDataSnapshots.add(index, snapshot);
+                notifyChangedListeners(OnChangedListener.EventType.Added, index);
+            } else {
+                mDataSnapshots.set(index, snapshot);
+                notifyChangedListeners(OnChangedListener.EventType.Changed, index);
+            }
+        } else {
+            Log.w(TAG, "Key not found at ref: " + snapshot.getRef());
+            if (isMatch(index, key)) {
+                mDataSnapshots.remove(index);
+                notifyChangedListeners(OnChangedListener.EventType.Removed, index);
+            }
+        }
+    }
+
+    @Override
+    public void onChildChanged(DataSnapshot snapshot, String previousChildKey) {
+        super.setOnChangedListener(null);
+        super.onChildChanged(snapshot, previousChildKey);
+        super.setOnChangedListener(mListener);
```

Comment:
I would move my comment here because it's more relevant here now:

@SUPERCILEX There is always alternative way by using a flag to temporary avoid notifying the listener:

```java
private boolean mDoNotNotifyChangeListeners;

...
@Override
public void onChildChanged(DataSnapshot snapshot, String previousChildKey) {
	mDoNotNotifyChangeListeners = true;
	super.onChildChanged(snapshot, previousChildKey);
	mDoNotNotifyChangeListeners = false;
}
...	

@Override
protected void notifyChangedListeners(OnChangedListener.EventType type, int index, int oldIndex) {
	if (!mDoNotNotifyChangeListeners){
		super.notifyChangedListeners(type, index, oldIndex);
	}
}
```

Quick facts:
- There are no `super` calls here (but they do not really bother me)
- We can remove local `mListener`
- Remove overriding of `setOnChangedListener`
- Now `notifyChangedListeners` is overriden
- It introduces an extra variable. with local `mListener` it uses only something that already existed in `super` implementation.
- The end result is the same

Since IMO this is a matter of elegance and feeling we should ask @puf which solution is nicer in his opinion.